### PR TITLE
Issue 4195. Add type_literal_in_constant_pattern

### PR DESCRIPTION
**Pull request by scheglov**
Obviously, cannot be landed with dependency overrides, will clean up after initial review.

Fixes https://github.com/dart-lang/linter/issues/4195

### CODE REVIEW

File: lib/src/rules/no_constant_pattern_type_literal.dart [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+void f(Object? x) {
+  if (x case int _) {
+    print('int');
+  }
+}
+```
+
+''';
+
+class NoConstantPatternTypeLiteral extends LintRule {
```
Comment (pq) [inline lib/src/rules/no_constant_pattern_type_literal.dart:53]: I think I'd prefer `NoConstantPatternTypeLiterals` --- on balance we have more plural thank singular names.

File: lib/src/rules/no_constant_pattern_type_literal.dart [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  static const String lintName = 'no_constant_pattern_type_literal';
```
Comment (bwilkerson) [inline lib/src/rules/no_constant_pattern_type_literal.dart:54]: Or `type_literal_in_constant_pattern`?

There's always been some inconsistency between lints names describing what you ought to do and describing what you did wrong. All of the non-lint diagnostics describing what's wrong with the code, and I'm leaning in that direction for this lint, but it's a judgement call.

**Review by pq — APPROVED** [review]
Looks great!  I'll defer to @eernstg in case he has any questions about semantics but this handles the cases I'm aware of (and most concerned about), so cool!

File: lib/src/rules/no_constant_pattern_type_literal.dart [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+  static const LintCode matchType = LintCode(
+    lintName,
+    "Use '== TypeName ' instead of a type literal.",
```
Comment (bwilkerson) [inline lib/src/rules/no_constant_pattern_type_literal.dart:58]: nit: extra space after 'TypeName'

**Review by bwilkerson — APPROVED** [review]

File: lib/src/rules/no_constant_pattern_type_literal.dart (hunk 1 above)
Comment (scheglov) [reply lib/src/rules/no_constant_pattern_type_literal.dart:53]: With the comment from Brian below, I renamed it to `TypeLiteralInConstantPattern`, to be consistent with the `type_literal_in_constant_pattern` name. Is this a correct way to name lint rules?

File: lib/src/rules/no_constant_pattern_type_literal.dart (hunk 2 above)
Comment (scheglov) [reply lib/src/rules/no_constant_pattern_type_literal.dart:54]: Done.

File: lib/src/rules/no_constant_pattern_type_literal.dart (hunk 3 above)
Comment (scheglov) [reply lib/src/rules/no_constant_pattern_type_literal.dart:58]: Fixed.

File: lib/src/rules/no_constant_pattern_type_literal.dart (hunk 1 above)
Comment (bwilkerson) [reply lib/src/rules/no_constant_pattern_type_literal.dart]: Yes. It's easier to find the implementation if the class name and lint name match.

**Pull request merged**